lib/galaxy/model/store/discover.py

```diff
@@ -165,6 +172,86 @@ def create_dataset(
 
         return primary_data
 
+    def populate_collection_elements(self, collection, root_collection_builder, filenames, name=None, metadata_source_name=None):
+        # TODO: allow configurable sorting.
+        #    <sort by="lexical" /> <!-- default -->
+        #    <sort by="reverse_lexical" />
+        #    <sort regex="example.(\d+).fastq" by="1:numerical" />
+        #    <sort regex="part_(\d+)_sample_([^_]+).fastq" by="2:lexical,1:numerical" />
+        if name is None:
+            name = "unnamed output"
+
+        element_datasets = []
+        for filename, discovered_file in filenames.items():
+            create_dataset_timer = ExecutionTimer()
+            fields_match = discovered_file.match
+            if not fields_match:
+                raise Exception("Problem parsing metadata fields for file %s" % filename)
+            element_identifiers = fields_match.element_identifiers
+            designation = fields_match.designation
+            visible = fields_match.visible
+            ext = fields_match.ext
+            dbkey = fields_match.dbkey
+            # galaxy.tools.parser.output_collection_def.INPUT_DBKEY_TOKEN
```

Comment:
Might be good to place this in a common place -- maybe `output_collection_def` belongs in `model.dataset_collections` ?

Reply:
Yeah, I'm not happy about that refactoring but I don't think tool output definition stuff belongs in the model layer. I think I pushed too much "job stuff" into the model discovery layer here... I think the fix is to be a little more fine grain but I'm not sure I can see it exactly and I'm not sure it is worth the time. I did definitely catch on this issue though - as I keep working on refactoring this stuff I'll try to fix the problem.